url7f15a160d2_timofei7_old_ai_assignments_tgz-pJ8-org_json_TestJ8/src/hw4/HW4Main.java

```diff
@@ -0,0 +1,318 @@
+		}
+		
+		//if just outputing cnf do that here and quit
+		if (algo == 4)
+		{
+			csp.outputCNF();
+			return true;
+		}
+		
+		//make first partial assignment
+		PartialAssignment pa = new PartialAssignment(cp.size);
+		
+		long start = System.currentTimeMillis();
+		
+		switch (algo)
+		{
+			case 1:
+				System.out.println("running with no heuristics..");
+				found = csp.backtrackingSearch(pa);
+				printSolution(found.assignment, cp);
+				break;
+			case 2:
+				System.out.println("running minimum remaining variable..");
+				found = csp.backtrackingSearchMRV(pa, dl);
+				printSolution(found.assignment, cp);
+				break;
+			case 3:
+				System.out.println("running minimum remaining variable with least constrainting value..");
+				found = csp.backtrackingSearchMRVLCV(pa, dl);
+				printSolution(found.assignment, cp);
+				break;
+			default:
+				System.out.println("invalid choice try again...");
+				break;
+		}
+		System.out.println("BACKTRACK COUNT: " + csp.count);
+		long elapsed = System.currentTimeMillis() - start;
+		//System.out.println("ELAPSED TIME: " + elapsed + " milliseconds to run search");
+		return found.value;
+	}
+	
+	
+	/**
+	 * mapcolor test from devin
+	 */
+	public static void test() {
+		
+		String variables = "WA.NT.SA.NSW.Q.V";
+		String values="r.g.b";
+		String constraint = "r.g:r.b:g.r:g.b:b.r:b.g";
+		
+		
+		CSP mapcolor = new CSP(variables, values);
+		
+		mapcolor.addConstraint("WA", "NT", constraint);
+		mapcolor.addConstraint("SA", "NT", constraint);
+		mapcolor.addConstraint("SA", "WA", constraint);
+		mapcolor.addConstraint("SA", "NSW", constraint);
+		mapcolor.addConstraint("SA", "V", constraint);
+		mapcolor.addConstraint("SA", "Q", constraint);
+		mapcolor.addConstraint("NSW", "Q", constraint);
+		mapcolor.addConstraint("NSW", "V", constraint);
+		
+		PartialAssignment pa = new PartialAssignment(6);
+		
+		mapcolor.backtrackingSearch(pa);
+	}
+	
+	
+	/**
+	 * use reflection to pass a converter to print the domainlist
+	 * mostly just my effort to learn a bit about java reflection
+	 */
+	@SuppressWarnings("unchecked")
+	public static void printDL(DomainList dl, CircuitProblem cp)
+	{
+		try
+		{
+			Class cls = Class.forName("hw4.Rect");
+			Object foo = cls.newInstance();
+			Method method = cls.getMethod("intToString", new Class[] { int.class, int.class });
+			System.out.println(dl.toString(foo, method, cp.width));
+		}
+		catch (Exception e)
+		{ 
+			System.out.println("error printing: " + e);
+		}
+	}
+	
+	/**
+	 * takes a start rect to indicate left upper corner coordinates
+	 * and then given the rect r builds a list of all absolute internal coordinates
+	 * @param orig
+	 * @param cols
+	 * @param r
+	 * @return set of coordinates
+	 */
+	public static Set<Integer> coordSet(int orig, int cols, Rect r)
+	{
+		Rect o = Rect.fromInt(orig, cols);
+		int xorg = o.width;
+		int yorg = o.height;
+		
+		Set<Integer> s = new HashSet<Integer>();
+				
+		for (int x=xorg; x < (xorg + r.width); x++)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;r&#39; (index: 2 | type: Rect). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;cols&#39;, &#39;orig&#39;. The same parameter type is validated here for example: url7f15a160d2_timofei7_old_ai_assignments_tgz-pJ8-org_json_TestJ8/src/hw4/CircuitProblem.java:36. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).